wespeaker/utils/executor.py

```diff
@@ -17,39 +17,46 @@
 
 import torch
 import torchnet as tnt
+from wespeaker.dataset.dataset_utils import apply_cmvn, spec_aug
 
 
-def run_epoch(dataloader,
-              epoch_iter,
-              model,
-              criterion,
-              optimizer,
-              scheduler,
-              margin_scheduler,
-              epoch,
-              logger,
-              scaler,
-              enable_amp,
-              log_batch_interval=100,
-              device=torch.device('cuda')):
+def run_epoch(dataloader, epoch_iter, model, criterion, optimizer, scheduler,
+              margin_scheduler, epoch, logger, scaler, device, configs):
     model.train()
     # By default use average pooling
     loss_meter = tnt.meter.AverageValueMeter()
     acc_meter = tnt.meter.ClassErrorMeter(accuracy=True)
 
+    frontend_type = configs['dataset_args'].get('frontend', 'fbank')
     for i, batch in enumerate(dataloader):
-        utts = batch['key']
-        targets = batch['label']
-        features = batch['feat']
-
         cur_iter = (epoch - 1) * epoch_iter + i
         scheduler.step(cur_iter)
         margin_scheduler.step(cur_iter)
 
-        features = features.float().to(device)  # (B,T,F)
-        targets = targets.long().to(device)
+        utts = batch['key']
+        targets = batch['label']
+        targets = targets.long().to(device)  # (B)
+        if frontend_type == 'fbank':
+            features = batch['feat']  # (B,T,F)
+            features = features.float().to(device)
+        else:  # 's3prl'
+            wavs = batch['wav']  # (B,1,W)
+            wavs = wavs.squeeze(1).float().to(device)  # (B,W)
+            wavs_len = torch.LongTensor([wavs.shape[1]]).repeat(
+                wavs.shape[0]).to(device)  # (B)
+            with torch.cuda.amp.autocast(enabled=configs['enable_amp']):
+                features, _ = model.module.frontend(wavs, wavs_len)
+
+        with torch.cuda.amp.autocast(enabled=configs['enable_amp']):
```

Comment:
I don't think it is necessary to add amp context here. There is no pytorch model involved.

Reply:
solved